### DISCUSSION TITLE
Convert JS Promises to Ruby Promises

### DISCUSSION THREAD

**Discussion by hmdne** [opal/opal#2215] [General]
As far as I know, there's no way to convert from JS Promises to Opal Promises.

I'm developing an Opal library dealing with a lot of JS Promises and I made this monkey-patch to support a behaviour `Promise.new(js_promise)` and, if we wanted to convert what a promise resolves to: `Promise.new(js_promise) { |i| Native(i) }`.

```ruby
class Promise
  alias initialize_promise initialize

  def initialize(native={}, &block)
    if native?(native)
      initialize_promise
      block = proc { |i| i } unless block_given?
      xself = self # Unsure about the necessity of this...
      native.JS.then(proc { |i| xself.resolve(block.call(i)) }.to_n)
            .JS.catch(proc { |i| xself.reject(Native(i)) }.to_n)
    else
      initialize_promise(native)
    end
  end
end
```

I want to make it into a pull request, but first I want to know what other Opalists think about this.

**Comment by elia** [comment]
I'm not completely up-to-day with the latest on Promise implementations and the support we have from browsers, but I wouldn't object to re-writing `Promise` as:

```rb
class Promise < `Promise`
  …
end
```

and make it bridge the native version… but it can also be added as a monkey-patch that updates `new` and `to_n` within `stdlib/native`, see `Hash` as an example:

https://github.com/opal/opal/blob/77e3d491ff2816fbeecd804475e306018abe373a/stdlib/native.rb#L532-L535

**Comment by hmdne** [comment]
A question:

Does a syntax like: ```class Promise < `Promise` ``` automatically advance the new/existing JS objects? This would sound counterintuitive, but maybe not really.

**Comment by hmdne** [comment]
Ok, few code files read and analyzed later and I see that yes, ```class Promise < `Promise` ``` syntax does affect existing objects. Anyway, there's a slight difference between how Opal and JS promises work. Opal promises are objects that respond to all 4: `then`, `rescue` and `resolve`, `reject`. JS promises only respond to `then` and `catch` (for the record: also `finally`). So, how do they compare. This code is how Opal promises work:

```ruby
def resolve_after(time)
  promise = Promise.new
  Delay.new(@native, time) { promise.resolve }.start
  promise
end
```

An equivalent JS promise would work like this:
```js
function resolve_after(time) {
  return new Promise(function(resolve,reject) {
    setTimeout(resolve, time);
  });
}
```

I think this makes them incompatible by a principle. We could without a problem implement a JS promise in terms of an Opal one (but that's useless), but not vice versa. In a way, maybe that's not a problem, many times we get a Promise from JS side which "promises" to resolve a said promise. For those we would raise if "promise.resolve" is used. For Opal ones, we would override an initializer and save those two (anonymous function) arguments (to an anonymous function "promise resolver" given as an argument to a constructor) "resolve" and "reject" for when `promise.resolve` or `reject` is called. We would provide our own "promise resolver".

So I agree, that would be the sanest way to fix this interoperability issue.

**Comment by elia** [comment]
You're right, you see any objection against going with the `#to_n` solution as proposed above?

**Comment by hmdne** [comment]
Not at all. I wrote the sentence "I think this makes them incompatible by a principle" before having entirely understood the JS promises, but haven't removed it, because logically it was still true. That looks like the best idea.

**Comment by hmdne** [comment]
Btw. I don't know how to express it, because the comment above stems again from a slight misunderstanding of how things work in Opal-land. Of course adding the converter functions is the easy way and fully workable (it would be more like give-away than converting), but long-term bridging them (```Promise < `Promise` ```) sounds like a more robust idea. I even thought about moving such an implementation to corelib, but then I realized that, well, Ruby doesn't include a Promise class at all.

**Comment by hmdne** [comment]
See https://github.com/opal/opal/pull/2220 for the ongoing work on bridged promises.

**Comment by hmdne** [comment]
This is now part of Opal since Opal 1.2 as PromiseV2
